homeassistant/components/gocr/image_processing.py

```diff
@@ -0,0 +1,162 @@
+        self._camera_entity = camera_entity
+        if name is not None:
+            self._name = name
+        else:
+            self._name = ("GOCR {0}".format(
+                split_entity_id(camera_entity)[1])
+                if config[CONF_NAME] == '' else
+                config[CONF_NAME])
+        self._state = None
+        if hasattr('config', CONF_X_POSITION) and config[CONF_X_POSITION] is not None:
+            self.crop = (config[CONF_X_POSITION], config[CONF_Y_POSITION],
+                         config[CONF_X_POSITION] + config[CONF_WIDTH],
+                         config[CONF_Y_POSITION] + config[CONF_HEIGHT])
+        else:
+            self.crop = None
+        self.rotate = config[CONF_ROTATE]
+        self.negate = config[CONF_NEGATE]
+        if config.get(CONF_CHARS) is not None:
+            digits = ['-C', config.get(CONF_CHARS)]
+        else:
+            digits = []
+        if config.get(CONF_UNRECOGNIZED) is not None:
+            unrecognized = ['-u', config.get(CONF_UNRECOGNIZED)]
+        else:
+            unrecognized = []
+        if config[CONF_THRESHOLD] is not None:
+            threshold = ['-l', str(config[CONF_THRESHOLD])]
+        else:
+            threshold = []
+        if config[CONF_EXTRA_ARGUMENTS] is not None:
+            extra_arguments = shlex.split(config[CONF_EXTRA_ARGUMENTS])
+        else:
+            extra_arguments = []
+
+        self._command = ([config[CONF_GOCR_BIN],
+                         '-e', '/dev/null', '-f', 'UTF8'] +
+                         digits + unrecognized + threshold +
+                         extra_arguments)
+        _LOGGER.info("Command : %s -i <tmpfile>",
+                     shlex.quote(' '.join(self._command)))
+
+    @property
+    def device_class(self):
+        """Return the class of this device, from component DEVICE_CLASSES."""
+        return 'ocr'
+
+    @property
+    def camera_entity(self):
+        """Return camera entity id from process pictures."""
+        return self._camera_entity
+
+    @property
+    def name(self):
+        """Return the name of the image processor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the entity."""
+        return self._state
+
+    def process_image(self, image):
+        """Process the image."""
+        from PIL import Image
+        import PIL.ImageOps
+        from subprocess import run, PIPE
+        import io
+        from tempfile import NamedTemporaryFile
+
+        _LOGGER.debug("Start processing")
+
+        stream = io.BytesIO(image)
+        img = Image.open(stream)
+        if self.crop is not None:
+            img = img.crop(self.crop)
+        if self.negate:
+            img = PIL.ImageOps.invert(img)
+        if self.rotate is not None:
+            img = img.rotate(self.rotate, expand=1)
+        tmp = NamedTemporaryFile(suffix='.ppm')
+        self._state = None
+        try:
+            img.save(tmp.name)
```

Comment:
Looks like `gocr` also supports reading the input image via stdin (https://www.systutorials.com/docs/linux/man/1-gocr/). That would be much preferable over creating a new temporary file each time. You can do this with `subprocess.Popen` and the `.communicate` method on the returned object.